garage/tf/models/base.py

```diff
@@ -0,0 +1,296 @@
+"""Base model classes."""
+import abc
+from collections import namedtuple
+import warnings
+
+import tensorflow as tf
+
+
+class BaseModel(abc.ABC):
+    """
+    Interface-only abstract class for models.
+
+    A Model contains the structure/configuration of a set of computation
+    graphs, or can be understood as a set of networks. Using a model
+    requires calling `build()` with given input placeholder, which can be
+    either tf.placeholder, or the output from another model. This makes
+    composition of complex models with simple models much easier.
+
+    Examples:
+        model = SimpleModel()
+        # To use a model, first create a placeholder.
+        # In the case of TensorFlow, we create a tf.placeholder.
+        input_ph = tf.placeholder(tf.float32)
+
+        # Building the model
+        output = model.build(input_ph)
+
+        # We can also pass the output of a model to another model.
+        # Here we pass the output from the above SimpleModel object.
+        model_2 = ComplexModel()
+        output_2 = model_2.build(output)
+
+    """
+
+    def build(self, *inputs):
+        """
+        Output of model with the given input placeholder(s).
+
+        This function is implemented by subclasses to create their computation
+        graphs, which will be managed by Model. Generally, subclasses should
+        implement `build()` directly.
+
+        Args:
+            inputs: Tensor input(s) for the model.
+
+        Return:
+            output: Tensor output(s) of the model.
+        """
+        pass
+
+    @property
+    def name(self):
+        """Name for this Model."""
+        pass
+
+    @property
+    def parameters(self):
+        """
+        Parameters of the Model.
+
+        The output of a model is determined by its parameter. It could be
+        the weights of a neural network model or parameters of a loss
+        function model.
+        """
+        pass
+
+    @parameters.setter
+    def parameters(self, parameters):
+        """Set parameters of the Model."""
+        pass
+
+
+class Network:
+    """
+    Network class For TensorFlow.
+
+    A Network contains connectivity information by inputs/outputs.
+    When a Network is built, it appears as a subgraph in the computation
+    graphs, scoped by the Network name. All Networks built with the same
+    model share the same parameters, i.e same inputs yield to same outputs.
+
+    Args:
+      model: The model building this network.
+      inputs: Input Tensor(s).
+      name: Name of the network, which is also the name scope.
+    """
+
+    def __init__(self, model, inputs, name):
+        self._inputs = inputs
+        with tf.name_scope(name=name):
```

Comment:
if we do this part inside model, you can remove the constructor dependence on model and not pass `self` (which is an anti-pattern).